contracts/SafeProtocolRegistry.sol

```diff
@@ -38,16 +40,33 @@ contract SafeProtocolRegistry is ISafeProtocolRegistry, Ownable2Step {
 
     /**
      * @notice Allows only owner to add a integration. A integration can be any address including zero address for now.
-     *         This function does not permit adding a integration twice.
-     *         TODO: Add logic to validate if integration implements correct interface.
+     *         This function does not permit adding a integration twice. This function validates if integration supports expected interfaceId.
      * @param integration Address of the integration
+     * @param integrationType Enum.IntegrationType indicating the type of integration
      */
-    function addIntegration(address integration, Enum.IntegrationType integrationType) external onlyOwner {
+    function addIntegration(address integration, Enum.IntegrationType integrationType) external virtual onlyOwner {
         IntegrationInfo memory integrationInfo = listedIntegrations[integration];
 
         if (integrationInfo.listedAt != 0) {
             revert CannotAddIntegration(integration);
         }
+
+        // Check if integration supports expected interface
+        if (
+            integrationType == Enum.IntegrationType.Hooks && !IERC165(integration).supportsInterface(type(ISafeProtocolHooks).interfaceId)
+        ) {
+            revert IntegrationDoesNotSupportExpectedInterfaceId(integration, type(ISafeProtocolHooks).interfaceId);
+        } else if (
+            integrationType == Enum.IntegrationType.Plugin && !IERC165(integration).supportsInterface(type(ISafeProtocolPlugin).interfaceId)
+        ) {
+            revert IntegrationDoesNotSupportExpectedInterfaceId(integration, type(ISafeProtocolPlugin).interfaceId);
+        } else if (
+            integrationType == Enum.IntegrationType.FunctionHandler &&
+            !IERC165(integration).supportsInterface(type(ISafeProtocolFunctionHandler).interfaceId)
+        ) {
+            revert IntegrationDoesNotSupportExpectedInterfaceId(integration, type(ISafeProtocolFunctionHandler).interfaceId);
+        }
```

Comment:
Perhaps merging this into a single if statement would yield better bytecode size/gas results. Would experiment with this if you have time

Reply:
1. For `addIntegration(...)` there is not much difference and in-fact current implementation (`a`) uses less gas.
2. For deployment of `SafeProtocolRegistry`, `b` uses less gas.

a. With current changes:
<img width="981" alt="Screenshot 2023-08-01 at 18 01 22" src="https://github.com/5afe/safe-core-protocol/assets/16358982/4a4f5882-5f8e-4bc1-9eb3-ebf7c8c6e633">



b. With single if statement [code](https://github.com/5afe/safe-core-protocol/commit/18ad58d632b859d82fbc6042cb1f4212a9c90a69)

<img width="978" alt="Screenshot 2023-08-01 at 18 01 56" src="https://github.com/5afe/safe-core-protocol/assets/16358982/4ccb3b7d-da25-427d-a7bd-4b03b5715d42">

Reply:
Keeping the current implementation because it would give information about what interface id is expected in the revert and gas usage difference is not significant while adding registry. Regarding, deployment cost, as this contract would be deployed only once, I think it is acceptable for now. We can reiterate on this if needed